### PR TITLE
 Rework add drives mechanism in svirt 

### PR DESCRIPTION
**Pull request by mloviska**
SLEM 5.1 builds bootable, ready to use images for z/KVM. To make sure
that images can be expanded we need to add additional functionality to
`add_disk` to create a backingfile of given size.
For `svirt` backend ( except of hyperv ), we should be able to create
drives in following modes.

1. Create empty drive
2. Create a backing image file of given size
3. Re-use given image
4. Add cd-rom drive with given iso image

### CODE REVIEW

File: consoles/sshVirtsh.pm [hunk 1]
```diff
@@ -325,111 +325,169 @@ sub add_interface {
...
 
+sub to_byte {
+    my $sp              = shift;
+    my $size_multiplier = {
+        b => 1,
+        k => 1024, K => 1024,
+        M => 1024 * 1024,
+        G => 1024 * 1024 * 1024,
+        T => 1024 * 1024 * 1024 * 1024,
+    };
```
Comment (okurz) [inline consoles/sshVirtsh.pm:336]: https://metacpan.org/pod/Number::Format might be more helpful here. There is also https://metacpan.org/pod/Number::Bytes::Human which is currently not in openSUSE:Factory but we could submit that

File: consoles/sshVirtsh.pm [hunk 2]
```diff
@@ -325,111 +325,169 @@ sub add_interface {
...
+    $dest .= basename($src);
+    die "Downloading $src failed!" if $self->run_cmd("rsync -av $src $dest");
+    return $dest;
+}
+
+sub decompress_cmd {
+    return sprintf(q(xz --decompress --keep --verbose --no-warn --quiet '%s'), shift);
+}
+
 sub add_disk {
```
Comment (okurz) [inline consoles/sshVirtsh.pm:401]: The method is quite big and complex. Could you please split the method in a separate commit before changing it? If you are unsure how to do that I could try myself to suggest a change.

**Review by okurz — CHANGES_REQUESTED** [review]

File: consoles/sshVirtsh.pm [hunk 3]
```diff
@@ -325,111 +325,169 @@ sub add_interface {
...
 }
 
+sub to_byte {
+    my $sp              = shift;
+    my $size_multiplier = {
+        b => 1,
+        k => 1024, K => 1024,
+        M => 1024 * 1024,
+        G => 1024 * 1024 * 1024,
+        T => 1024 * 1024 * 1024 * 1024,
```
Comment (Martchus) [inline consoles/sshVirtsh.pm:335]: In Perl there's a operator for computing an exponentiation, e.g.:
```suggestion
        T => 1024 ** 3,
```
But `perl-Number-Format` is already in Factory so you might just depend on that.

File: consoles/sshVirtsh.pm [hunk 4]
```diff
@@ -325,111 +325,169 @@ sub add_interface {
     return;
 }
 
+sub to_byte {
+    my $sp              = shift;
```
Comment (Martchus) [inline consoles/sshVirtsh.pm:329]: It would be nice to use signatures in new functions, e.g.:
```suggestion
sub to_byte ($sp) {
```

File: consoles/sshVirtsh.pm [hunk 5]
```diff
@@ -325,111 +325,169 @@ sub add_interface {
...
+    my $size_multiplier = {
+        b => 1,
+        k => 1024, K => 1024,
+        M => 1024 * 1024,
+        G => 1024 * 1024 * 1024,
+        T => 1024 * 1024 * 1024 * 1024,
+    };
+
+    #in case of invalid format return error
```
Comment (Martchus) [inline consoles/sshVirtsh.pm:338]:
```suggestion
    # in case of invalid format return error
```

**Review by Martchus — CHANGES_REQUESTED** [review]

File: consoles/sshVirtsh.pm [hunk 6]
```diff
@@ -325,111 +325,169 @@ sub add_interface {
...
+        return undef;
+    }
+
+    return $+{num} * $size_multiplier->{$+{unit} // 'b'};
+}
+
+sub gen_kvm_img {
+    my ($self, %args) = @_;
+    my $byte_size = to_byte($args{size});
+    die('Missing argument size or invalid format') unless (defined($byte_size));
```
Comment (Martchus) [inline consoles/sshVirtsh.pm:349]: I would generally prefer to use less parentheses like this:
```suggestion
    die 'Missing argument size or invalid format' unless defined $byte_size;
```

File: consoles/sshVirtsh.pm [hunk 7]
```diff
@@ -325,111 +325,169 @@ sub add_interface {
...
+    for (1 .. $tries) {
+        ($ret, $stdout, $stderr) = $self->run_cmd(join(" ", @cmd), wantarray => 1);
+        if ($stderr) {
+            bmwqemu::diag("gen_kvm_img failed: $stderr, try $_ out of $tries");
+            sleep 5;
+        }
+        last unless $ret;
+    }
+    die "Too many attempts to format HDD" if $stderr;
+    bmwqemu::diag("gen_kvm_img: Finished creating image file...");
```
Comment (Martchus) [inline consoles/sshVirtsh.pm:378]: If no interpolation is required I'd prefer to use single quotes.

File: consoles/sshVirtsh.pm [hunk 8]
```diff
@@ -325,111 +325,169 @@ sub add_interface {
...
+        my $basedir     = "/vmfs/volumes/$vmware_datastore/openQA/";
+        my $nfs_basedir = '/vmfs/volumes/' . get_required_var('VMWARE_NFS_DATASTORE');
+        $file = sprintf("%s%s.vmdk", $self->{name}, $args->{dev_id});
+        my $vmware_disk_path = $basedir . $file;
+
+        # Power VM off, delete it's disk image, and create it again.
+        # Than wait for some time for the VM to *really* turn off.
+        my $cmd = <<"EOT";
+( set -x; vmid=\$(vim-cmd vmsvc/getallvms | awk '/$self->{name}/ { print \$1 }');
+if [ \$vmid ]; then
```
Comment (Martchus) [inline consoles/sshVirtsh.pm:423]: Is it actually required to escape the `$` here?

Comment (mloviska) [reply consoles/sshVirtsh.pm:423]: Yes, that should be treated as a shell variable.

File: consoles/sshVirtsh.pm (hunk 4 above)
Comment (mloviska) [reply consoles/sshVirtsh.pm:329]: Yeah, sure. But wasn't there a reason behind to avoid signatures in Perl?

Comment (Martchus) [reply consoles/sshVirtsh.pm:329]: Compatibility - but at this point we decided to drop support for too old Perl versions.

File: consoles/sshVirtsh.pm (hunk 8 above)
Comment (Martchus) [reply consoles/sshVirtsh.pm:423]: You're right, apparently interpolation is happening here. (`<<'EOT'` would avoid it but it looks like you actually need it here.)

File: consoles/sshVirtsh.pm (hunk 1 above)
Comment (mloviska) [reply consoles/sshVirtsh.pm]: > https://metacpan.org/pod/Number::Format might be more helpful here. There is also https://metacpan.org/pod/Number::Bytes::Human which is currently not in openSUSE:Factory but we could submit that

`Number::Format` seems to work nicely, thanks!

File: consoles/sshVirtsh.pm (hunk 4 above)
Comment (okurz) [reply consoles/sshVirtsh.pm]: yes, years ago. But to my understanding signatures are fine as long as we avoid some more exotic features.

File: t/22-svirt.t [hunk 9]
```diff
@@ -234,4 +235,479 @@ subtest 'Method backend::svirt::open_serial_console_via_ssh()' => sub {
...
+            push @last_ssh_args,     [%args];
+
+            my $ret = shift @ssh_cmd_return;
+            return undef unless defined $ret;
+
+            if (ref($ret) eq 'ARRAY') {
+                return @$ret;
+            } else {
+                return $ret;
+            }
```
Comment (okurz) [inline t/22-svirt.t:258]:
```suggestion
            return ref($ret) eq 'ARRAY' ? @$ret : $ret;
```

File: t/22-svirt.t [hunk 10]
```diff
@@ -234,4 +235,479 @@ subtest 'Method backend::svirt::open_serial_console_via_ssh()' => sub {
...
+            is(defined($source_node),              1,                    '<source> node exists');
+            is($source_node->getAttribute('file'), $basedir . $exp_file, 'file attribute of <source> is correct');
+
+            is(defined($driver_node),               1,        '<driver> node exists');
+            is($driver_node->getAttribute('name'),  'qemu',   'name attribute of <driver> is correct');
+            is($driver_node->getAttribute('type'),  'qcow2',  'type attribute of <driver> is correct');
+            is($driver_node->getAttribute('cache'), 'unsafe', 'cache attribute of <driver> is correct');
+        };
+
+
```
Comment (okurz) [inline t/22-svirt.t:466]: delete one blank line

File: t/22-svirt.t [hunk 11]
```diff
@@ -234,4 +235,479 @@ subtest 'Method backend::svirt::open_serial_console_via_ssh()' => sub {
...
+        subtest 'family svirt-xen-hvm backingfile=1' => sub {
+
```
Comment (okurz) [inline t/22-svirt.t:468]: delete blank line

File: t/22-svirt.t [hunk 12]
```diff
@@ -234,4 +235,479 @@ subtest 'Method backend::svirt::open_serial_console_via_ssh()' => sub {
...
+                    backingfile => 1,
+                    dev_id      => $dev_id,
+                    file        => '/my/path/to/this/file/' . $file,
+                    size        => 12
+            });
+            like($last_ssh_commands[0], qr%^rsync.*/my/path/to/this/file/$file $basedir$file%, 'Use rsync to copy file');
+            is($last_ssh_commands[-1], "qemu-img create -f qcow2 -b $basedir$file ${basedir}openQA-SUT-1$dev_id.img 12G", 'Used image size > backingfile size');
+        };
+
+
```
Comment (okurz) [inline t/22-svirt.t:484]: delete one blank line

File: t/22-svirt.t [hunk 13]
```diff
@@ -234,4 +235,479 @@ subtest 'Method backend::svirt::open_serial_console_via_ssh()' => sub {
...
+            $svirt->add_disk({
+                    backingfile => 1,
+                    dev_id      => $dev_id,
+                    file        => '/my/path/to/this/file/' . $file,
+                    size        => 5
+            });
+            like($last_ssh_commands[0], qr%^rsync.*/my/path/to/this/file/$file $basedir$file%, 'Use rsync to copy file');
+            is($last_ssh_commands[-1], "qemu-img create -f qcow2 -b $basedir$file ${basedir}openQA-SUT-1$dev_id.img $_10gb", 'Used image size <= backingfile size');
+
+
```
Comment (okurz) [inline t/22-svirt.t:500]: delete one blank line

File: t/22-svirt.t [hunk 14]
```diff
@@ -234,4 +235,479 @@ subtest 'Method backend::svirt::open_serial_console_via_ssh()' => sub {
...
+            is($source_node->getAttribute('file'), $basedir . "openQA-SUT-1$dev_id.img", 'file attribute of <source> is correct');
+
+            is(defined($driver_node),               1,        '<driver> node exists');
+            is($driver_node->getAttribute('name'),  'qemu',   'name attribute of <driver> is correct');
+            is($driver_node->getAttribute('type'),  'qcow2',  'type attribute of <driver> is correct');
+            is($driver_node->getAttribute('cache'), 'unsafe', 'cache attribute of <driver> is correct');
+
+            throws_ok { $svirt->add_disk({backingfile => 1, dev_id => $dev_id}) } qr/file/, "Die on missing file argument";
+        };
+        subtest 'family svirt-xen-hvm cdrom=1' => sub {
```
Comment (okurz) [inline t/22-svirt.t:518]: add one blank line for separation

File: t/22-svirt.t [hunk 15]
```diff
@@ -234,4 +235,479 @@ subtest 'Method backend::svirt::open_serial_console_via_ssh()' => sub {
...
+
+            is(defined($source_node),              1,                '<source> node exists');
+            is($source_node->getAttribute('file'), $basedir . $file, 'file attribute of <source> is correct');
+
+            is(defined($driver_node),               1,      '<driver> node exists');
+            is($driver_node->getAttribute('name'),  'qemu', 'name attribute of <driver> is correct');
+            is($driver_node->getAttribute('type'),  'raw',  'type attribute of <driver> is correct');
+            is($driver_node->getAttribute('cache'), undef,  'cache attribute of <driver> is not defined');
+
+
```
Comment (okurz) [inline t/22-svirt.t:545]: delete one blank line

File: t/22-svirt.t [hunk 16]
```diff
@@ -234,4 +235,479 @@ subtest 'Method backend::svirt::open_serial_console_via_ssh()' => sub {
...
+            throws_ok { $svirt->add_disk({cdrom => 1, dev_id => $dev_id}) } qr/file/, "Die on missing file argument";
+        };
+    };
+    subtest 'family kvm' => sub {
+
```
Comment (okurz) [inline t/22-svirt.t:550]: delete one blank line

File: t/22-svirt.t [hunk 17]
```diff
@@ -234,4 +235,479 @@ subtest 'Method backend::svirt::open_serial_console_via_ssh()' => sub {
...
+            }
+
+            my $dev_id   = 'dev_id_014_NO_SIZE';
+            my $exp_file = $svirt->name . $dev_id . ".img";
+            $svirt->add_disk({create => 1, dev_id => $dev_id});
+            is($last_ssh_commands[-1], "qemu-img create -f qcow2 $basedir$exp_file 20G", "Default size is 20G");
+        };
+
+        subtest 'family svirt-xen-hvm backingfile=1' => sub {
+
```
Comment (okurz) [inline t/22-svirt.t:598]: delete blank line

File: t/22-svirt.t [hunk 18]
```diff
@@ -234,4 +235,479 @@ subtest 'Method backend::svirt::open_serial_console_via_ssh()' => sub {
...
+                    cdrom     => 1,
+                    dev_id    => $dev_id,
+                    file      => '/my/path/to/this/file/' . $file,
+                    bootorder => 'MyArbitraryBootOrderArgument'
+            });
+            my $target_nodelist = $svirt->{domainxml}->findnodes('domain/devices/disk[@type="file" and @device="cdrom"]/boot[@order="MyArbitraryBootOrderArgument"]');
+            is($target_nodelist->size(), 1, 'Boot order entry was created');
+        }
+    };
+
```
Comment (okurz) [inline t/22-svirt.t:711]: delete blank line

File: t/22-svirt.t (hunk 9 above)
Comment (mloviska) [reply t/22-svirt.t:258]: I find more readable the multiline part. Sorry, I would prefer to keep it that way. The same applies above.

File: consoles/sshVirtsh.pm (hunk 2 above)
Comment (mloviska) [reply consoles/sshVirtsh.pm:395]: Yes, please, it would be nice if you can try to split the original method. Thanks

File: consoles/sshVirtsh.pm (hunk 4 above)
Comment (mloviska) [reply consoles/sshVirtsh.pm]: Maybe I am doing something wrong, but is if enable signatures with `use feature 'signatures'` I get warnings the feature is experimental

Comment (Martchus) [reply consoles/sshVirtsh.pm]: I would recommend enabling them via `use Mojo::Base -strict, -signatures;` (or just add `, -signatures` to the `use` args where `Mojo::Base` is already used).

Comment (mloviska) [reply consoles/sshVirtsh.pm]: > I would recommend enabling them via `use Mojo::Base -strict, -signatures;` (or just add `, -signatures` to the `use` args where `Mojo::Base` is already used).

Thanks, this was the magic. Good to know how to enable signatures in Perl. :)

File: consoles/sshVirtsh.pm (hunk 2 above)
Comment (okurz) [reply consoles/sshVirtsh.pm:390]: I did so in https://github.com/os-autoinst/os-autoinst/pull/1752 . I am sure this will cause multiple conflicts for you and I have only tested it with t/22-svirt.t so feel free to incorporate that in anyways helpful to you and please test accordingly.

**Comment by okurz** [comment]
now please rebase
